File: python/bank-account/bank_account.py
from enum import Enum
from threading import Lock


class BankAccount:
    State = Enum("State", ("new", "open", "closed"))

    def __init__(self):
        self._balance = 0
        self._state = BankAccount.State.new
        self._lock = Lock()

    def get_balance(self):
        with self._lock:
            if self._state == BankAccount.State.closed:
                raise ValueError(f"Can't get balance of a closed account.")
            return self._balance

    def open(self):
        with self._lock:
            if self._state == BankAccount.State.open:
                raise ValueError(f"Can't open already opened account.")
            self._state = BankAccount.State.open

    def deposit(self, amount):
        with self._lock:
            if amount < 0:
                raise ValueError(f"Can't deposit negative amount: {amount}.")
            if self._state == BankAccount.State.closed:
                raise ValueError(f"Can't deposit to a closed accocunt.")
            self._balance += amount

    def withdraw(self, amount):
        with self._lock:
            if amount < 0:
                raise ValueError(f"Can't withdraw negative amount: {amount}.")
            if self._state == BankAccount.State.closed:
                raise ValueError(f"Can't withdraw from a closed acocunt.")
            if amount > self._balance:
                raise ValueError(f"Can't withdraw more than current balance.")
            self._balance -= amount

    def close(self):
        with self._lock:
            if self._state == BankAccount.State.closed:
                raise ValueError(f"Can't close already closed account.")
            if self._state == BankAccount.State.new:
                raise ValueError(f"Can't close not opened account.")
            self._state = BankAccount.State.closed
            self._balance = 0


if __name__ == "__main__":
    pass

File: go/bank-account/bank_account.go
package account

import "sync"

// Account represents a bank account
type Account struct {
	balance int64
	closed  bool
	mux     sync.Mutex
}

// Open creates an account with an initial deposit
func Open(initialDeposit int64) *Account {
	if initialDeposit < 0 {
		return nil
	}
	return &Account{balance: initialDeposit}
}

// Close closes the account
func (a *Account) Close() (int64, bool) {
	a.mux.Lock()
	defer a.mux.Unlock()

	if a.closed {
		return 0, false
	}

	payout := a.balance
	a.balance = 0
	a.closed = true
	return payout, true
}

// Balance gives the current balance of the account
func (a *Account) Balance() (int64, bool) {
	a.mux.Lock()
	defer a.mux.Unlock()

	if a.closed {
		return 0, false
	}

	return a.balance, true
}

// Deposit adds/removes ammount from the account
func (a *Account) Deposit(amount int64) (int64, bool) {
	a.mux.Lock()
	defer a.mux.Unlock()

	if a.closed {
		return 0, false
	}

	if a.balance+amount < 0 {
		return 0, false
	}

	a.balance += amount
	return a.balance, true
}
